src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,159 @@
+#include "simeng/PerceptronPredictor.hh"
+
+namespace simeng {
+
+PerceptronPredictor::PerceptronPredictor(ryml::ConstNodeRef config)
+    : btbBits_(config["Branch-Predictor"]["BTB-Tag-Bits"].as<uint64_t>()),
+      globalHistoryLength_(
+          config["Branch-Predictor"]["Global-History-Length"].as<uint64_t>()),
+      rasSize_(config["Branch-Predictor"]["RAS-entries"].as<uint64_t>()) {
+  // Build BTB based on config options
+  uint32_t btbSize = (1 << btbBits_);
+  btb_.resize(btbSize);
+  // Initialise perceptron values with 0 for the global history weights, and 1 for the bias weight;
+  // and intialise the target with 0 (i.e., unknown)
+  for (int i = 0; i < btbSize; i++) {
+    btb_[i].first.assign(globalHistoryLength_, 0);
+    btb_[i].first.push_back(1);
+    btb_[i].second = 0;
+  }
+
+  // Set up training threshold according to empirically determined formula
+  trainingThreshold_ = (int)((1.93 * globalHistoryLength_) + 14);
+}
+
+PerceptronPredictor::~PerceptronPredictor() {
+  ras_.clear();
+  rasHistory_.clear();
+}
+
+BranchPrediction PerceptronPredictor::predict(uint64_t address, BranchType type,
+                                              int64_t knownOffset) {
+  // Get the hashed index for the prediction table.  XOR the global history with the
+  // non-zero bits of the address, and then keep only the btbBits_ bits of the output
+  // keep it in bounds of the prediction table.
```

Comment:
"and then keep only the btbBits_ bits of the output keep it in bounds of the prediction table." doesn't flow well. I think its missing a word or punctuation?

Reply:
"output **to** keep"?